### PR TITLE
feat(rslint_parser): Ban template literal after optional chain

### PR DESCRIPTION
**Pull request by Boshen**
## Summary
Template string is not allowed to be passed to tail position of optional chain.

### Test262 comparison coverage results on ubuntu-latest
| Test result | `main` count | This PR count | Difference |
| :---------: | :----------: | :-----------: | :--------: |
| Total | 44951 | 44951 | 0 |
| Passed | 43841 | 43845 | ✅ ⏫ **+4** |
| Failed | 1109 | 1105 | ✅ ⏬ **-4** |
| Panics | 1 | 1 | 0 |
| Coverage | 97.53% | 97.54% | **+0.01%** |

<details><summary><b>:tada: Fixed (4):</b></summary>

```
xtask/src/coverage/test262/test/language/expressions/optional-chaining/early-errors-tail-position-null-optchain-template-string-esi.js
xtask/src/coverage/test262/test/language/expressions/optional-chaining/early-errors-tail-position-null-optchain-template-string.js
xtask/src/coverage/test262/test/language/expressions/optional-chaining/early-errors-tail-position-optchain-template-string-esi.js
xtask/src/coverage/test262/test/language/expressions/optional-chaining/early-errors-tail-position-optchain-template-string.js
```
</details>

### CODE REVIEW

File: crates/rslint_parser/src/syntax/expr.rs [hunk 1]
```diff
@@ -724,15 +724,17 @@ fn subscripts(
     context: ExpressionContext,
 ) -> CompletedMarker {
     // test_err subscripts_err
-    // foo()?.baz[].
+    // foo()?.baz[].;
```
Comment (Boshen) [inline crates/rslint_parser/src/syntax/expr.rs:727]:
```
// test_err subscripts_err
// foo()?.baz[].
// BAR`b
```
This wrapped, and triggered the wrong error :-(

File: crates/rslint_parser/test_data/inline/err/template_after_optional_chain.rast [hunk 2]
```diff
@@ -0,0 +1,204 @@
...
+  │
+2 │ obj.val?.[expr]`template`
+  │ ^^^^^^^^^^^^^^^^^^^^^^^^^
+
+--
+error[SyntaxError]: Invalid tagged template on optional chain
+  ┌─ template_after_optional_chain.js:3:1
+  │
+3 │ obj.func?.(args)`template`
+  │ ^^^^^^^^^^^^^^^^^^^^^^^^^^
```
Comment (ematipico) [inline crates/rslint_parser/test_data/inline/err/template_after_optional_chain.rast:199]: Are we able to apply the range of the single template literal?

Comment (Boshen) [reply crates/rslint_parser/test_data/inline/err/template_after_optional_chain.rast:199]: This whole expression is called "Tagged Template", so using the whole range or the template range are both good. But the error message mentions "optional chain" so the whole range is better?

We can apply to the single template literal, but then we have to pass an extra argument to `parse_template_literal` 🤔

File: crates/rslint_parser/src/syntax/expr.rs [hunk 3]
```diff
@@ -797,7 +803,19 @@ fn subscripts(
...
-            BACKTICK => lhs = parse_template_literal(p, Present(lhs)),
+            BACKTICK => {
+                lhs = parse_template_literal(p, Present(lhs));
+                // test_err template_after_optional_chain
+                // obj.val?.prop`template`
+                // obj.val?.[expr]`template`
+                // obj.func?.(args)`template`
+                if is_optional_chain {
+                    p.error(
+                        p.err_builder("Invalid tagged template on optional chain")
```
Comment (MichaReiser) [inline crates/rslint_parser/src/syntax/expr.rs:814]: Can we use something closer to TS?

> Tagged template expressions are not permitted in an optional chain.

Where it becomes clear of whats

**Review by MichaReiser — APPROVED** [review]

File: crates/rslint_parser/src/syntax/expr.rs (hunk 3 above)
Comment (MichaReiser) [inline crates/rslint_parser/src/syntax/expr.rs:814]: Can we change the type of the whole expression to an `UNKNOWN_EXPRESSION`?

File: crates/rslint_parser/test_data/inline/err/template_after_optional_chain.rast (hunk 2 above)
Comment (MichaReiser) [reply crates/rslint_parser/test_data/inline/err/template_after_optional_chain.rast:199]: Nit: An option would be to capture the `lhs` range before calling into `parse_template_literal` and then add a secondary range to the diagnostic, pointing to the optional chain expression. But I'm not sure if it adds a lot of clarity or makes things even more confusing.

**Comment by Boshen** [comment]
I have updated the error message and changed it to unknown expression.

> Nit: An option would be to capture the lhs range before calling into parse_template_literal and then add a secondary range to the diagnostic, pointing to the optional chain expression. But I'm not sure if it adds a lot of clarity or makes things even more confusing.

Looks like it'll get more confusing :-(

**Review by ematipico — APPROVED** [review]

**Pull request merged**